src/target/riscv/riscv.c

```diff
@@ -462,11 +462,92 @@ static void trigger_from_breakpoint(struct trigger *trigger,
 	trigger->unique_id = breakpoint->unique_id;
 }
 
-static int maybe_add_trigger_t1(struct target *target,
-		struct trigger *trigger, uint64_t tdata1)
+static bool match_napot_checker(struct trigger *trigger, riscv_reg_t *tdata2)
+{
+	riscv_reg_t addr = trigger->address;
+	riscv_reg_t size = trigger->length;
+	if (size > 1 &&
+		(size & (size - 1)) == 0 &&
+		(addr & (size - 1)) == 0) {
+		if (tdata2)
+			*tdata2 = addr | ((size - 1) >> 1);
+		return true;
+	}
+	return false;
+}
+
+static int find_trigger(struct target *target, int type, int num, int *idx)
+{
+	RISCV_INFO(r);
+
+	riscv_reg_t tselect;
+	if (riscv_get_register(target, &tselect, GDB_REGNO_TSELECT) != ERROR_OK)
+		return ERROR_FAIL;
+
+	unsigned int i;
+	int found = 0, cnt = 0;
+	for (i = 0; i < r->trigger_count; i++) {
+		if (r->trigger_unique_id[i] == -1) {
+			riscv_reg_t tdata1 = 0;
+			riscv_set_register(target, GDB_REGNO_TSELECT, i);
+			if (riscv_get_register(target, &tdata1, GDB_REGNO_TDATA1) == ERROR_OK) {
+				int t = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+				if (type == t) {
+					found = (++cnt == num);
+					if (found)
+						break;
+					continue;
+				}
+			}
+		}
+		cnt = 0;
+	}
+
+	riscv_set_register(target, GDB_REGNO_TSELECT, tselect);
+
+	if (found) {
+		if (idx)
+			*idx = i - (num - 1);
+		return ERROR_OK;
+	}
+
+	return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+}
+
+static int set_trigget(struct target *target, int idx, riscv_reg_t tdata1, riscv_reg_t tdata2)
+{
+	int ret = ERROR_OK;
+	riscv_reg_t tselect, tdata1_rb;
+	if (riscv_get_register(target, &tselect, GDB_REGNO_TSELECT) != ERROR_OK)
+		return ERROR_FAIL;
+
+	riscv_set_register(target, GDB_REGNO_TSELECT, idx);
+	do {
+		riscv_set_register(target, GDB_REGNO_TDATA1, tdata1);
+		riscv_get_register(target, &tdata1_rb, GDB_REGNO_TDATA1);
+		if (tdata1 != tdata1_rb) {
+			LOG_DEBUG("Trigger doesn't support what we need; After writing 0x%"
+				PRIx64 " to tdata1 it contains 0x%" PRIx64,
+				tdata1, tdata1_rb);
+			riscv_set_register(target, GDB_REGNO_TDATA1, 0);
+			ret = ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+			break;
+		}
+		riscv_set_register(target, GDB_REGNO_TDATA1, tdata2);
+	} while (0);
+	riscv_set_register(target, GDB_REGNO_TSELECT, tselect);
+
+	return ret;
+}
+
+static int maybe_add_trigger_t1(struct target *target, struct trigger *trigger)
```

Comment:
There's no need to modify this function. Trigger type t1 is a legacy trigger (per the debug spec). It is only implemented by the HiFive1 AFAIK. The code is here to support existing functionality, but there's no need to add anything new. (Unless you're working with hardware that supports type 1 and some of the new ones.)